packages/agreement-process/src/services/agreementService.ts

```diff
@@ -324,17 +577,56 @@ export function agreementServiceBuilder(
       { authData, correlationId, logger }: WithLogger<AppContext>
     ): Promise<Agreement> {
       logger.info(`Suspending agreement ${agreementId}`);
-      const [agreement, events] = await suspendAgreementLogic({
+
+      const agreement = await retrieveAgreement(agreementId, agreementQuery);
+
+      assertRequesterIsConsumerOrProducer(agreement.data, authData);
+
+      assertExpectedState(
         agreementId,
+        agreement.data.state,
+        agreementSuspendableStates
+      );
+
+      const eservice = await retrieveEService(
+        agreement.data.eserviceId,
+        eserviceQuery
+      );
+
+      const descriptor = eservice.descriptors.find(
+        (d) => d.id === agreement.data.descriptorId
+      );
+      assertDescriptorExist(
+        eservice.id,
+        agreement.data.descriptorId,
+        descriptor
+      );
+
+      const updatedAgreement: Agreement = createAgreementSuspended({
+        agreement: agreement.data,
         authData,
-        agreementQuery,
-        tenantQuery,
-        eserviceQuery,
-        correlationId,
+        descriptor,
+        consumer: await retrieveTenant(agreement.data.consumerId, tenantQuery),
       });
-      await repository.createEvent(events);
 
-      return agreement;
+      const isProducer = authData.organizationId === agreement.data.producerId;
+      const isConsumer = authData.organizationId === agreement.data.consumerId;
+
+      if (!isProducer && !isConsumer) {
+        throw new Error(
+          "Agreement can only be suspended by the consumer or producer."
+        );
+      }
```

Comment:
This shall not be needed, we already call `assertRequesterIsConsumerOrProducer` above

Reply:
> This shall not be needed, we already call `assertRequesterIsConsumerOrProducer` above

Right, thanks, I would still need to use one of those check when creating the eventType, I'll change it soon!

Reply:
True, you could `throw genericError` instead of `new Error`, so that we use the errors we defined in `models` (even though this error will never happen...)